File: Daily/PY/Leetcode1584-连接所有点的最小费用.py
from typing import List
class Solution:
    def minCostConnectPoints(self, points: List[List[int]]) -> int:
        n = len(points)
        parent = list(range(n))
        def find(x):
            if parent[x] == x:
                return x
            parent[x] = find(parent[x])
            return parent[x]
        def union(x,y)->bool:
            fx = find(x)
            fy = find(y)
            if fx == fy:
                return False
            parent[fy] = fx
            return True
        
        def manhattan(x,y)->int:
            return abs(points[x][0] - points[y][0]) + abs(points[x][1] - points[y][1])
        #创建空list
        edges = list()

        for i in range(n):
            for j in range(i + 1,n):
                edges.append((manhattan(i,j),i,j))
        #py的sort
        edges.sort()
        ans , num = 0,1
        for arg1,arg2,arg3 in edges:
            if union(arg2,arg3):
                ans += arg1
                num += 1
                if num == n:
                    break
        return ans

File: Daily/sbgo/Leetcode1584-连接所有点的最小费用.go
package sbgo

import "sort"

func minCostConnectPoints(points [][]int) int {
	n := len(points)
	parent := make([]int, n)
	for i := range parent {
		parent[i] = i
	}
	var find func(x int) int
	find = func(x int) int {
		if x == parent[x] {
			return x
		}
		parent[x] = find(parent[x])
		return parent[x]
	}
	union := func(x, y int) bool {
		fx, fy := find(x), find(y)
		if fx == fy {
			return false
		}
		parent[fy] = fx
		return true
	}
	manhattan := func(x, y int) int {
		return abs(points[x][0]-points[y][0]) + abs(points[x][1]-points[y][1])
	}
	type edge struct{ x, y, len int }
	edges := []edge{}
	for i := range points {
		for j := i + 1; j < n; j++ {
			edges = append(edges, edge{i, j, manhattan(i, j)})
		}
	}
	sort.Slice(edges, func(i, j int) bool { return edges[i].len < edges[j].len })
	ans := 0
	num := 1
	for _, eg := range edges {
		if union(eg.x, eg.y) {
			ans += eg.len
			num++
			if num == n {
				break
			}
		}
	}
	return ans
}

func abs(x int) int {
	if x > 0 {
		return x
	}
	return -x
}
